libyul/backends/evm/ssa/Stack.h

```diff
@@ -0,0 +1,309 @@
+	friend constexpr auto operator<=>(size_t lhs, Offset rhs) noexcept { return lhs <=> rhs.value; }
+	friend constexpr bool operator==(Offset lhs, size_t rhs) noexcept { return lhs.value == rhs; }
+	friend constexpr bool operator==(size_t lhs, Offset rhs) noexcept { return lhs == rhs.value; }
+
+	/// Distance from the stack top (depth 0 = top).
+	/// Natural for stack operations (SWAP1 = swap with depth 1); used for operations that
+	/// conceptually work "from the top".
+	struct Depth
+	{
+		explicit constexpr Depth(size_t _value) : value(_value) {}
+		size_t value;
+		auto operator<=>(Depth const&) const = default;
+	};
+	// comparison operations with size_t
+	friend constexpr auto operator<=>(Depth lhs, size_t rhs) noexcept { return lhs.value <=> rhs; }
+	friend constexpr auto operator<=>(size_t lhs, Depth rhs) noexcept { return lhs <=> rhs.value; }
+	friend constexpr bool operator==(Depth lhs, size_t rhs) noexcept { return lhs.value == rhs; }
+	friend constexpr bool operator==(size_t lhs, Depth rhs) noexcept { return lhs == rhs.value; }
+
+	Stack(
+		Data& _data,
+		Callbacks _callbacks
+	):
+		m_data(&_data),
+		m_callbacks(std::move(_callbacks))
+	{}
+
+	Slot const& top() const
+	{
+		yulAssert(!m_data->empty());
+		return m_data->back();
+	}
+
+	void swap(Depth const& _depth)
+	{
+		yulAssert(swapReachable(_depth), "Stack too deep");
+		std::swap((*m_data)[depthToOffset(_depth).value], m_data->back());
+		if constexpr (!std::is_same_v<Callbacks, NoOpStackManipulationCallbacks>)
+			m_callbacks.swap(_depth.value);
+	}
+	void swap(Offset const& _offset) { swap(offsetToDepth(_offset)); }
+
+	/// if the stack state needs to be updated without notifying the callback, the template parameter can be set to false
+	template<bool callback=true>
+	void pop()
+	{
+		yulAssert(!m_data->empty());
+		m_data->pop_back();
+		if constexpr (callback && !std::is_same_v<Callbacks, NoOpStackManipulationCallbacks>)
+			m_callbacks.pop();
+	}
+
+	/// if the stack state needs to be updated without notifying the callback, the template parameter can be set to false
+	template<bool callback=true>
+	void push(Slot const& _slot)
+	{
+		m_data->emplace_back(_slot);
+		if constexpr (callback && !std::is_same_v<Callbacks, NoOpStackManipulationCallbacks>)
+			m_callbacks.push(_slot);
+	}
+
+	void dup(Depth const& _depth)
+	{
+		yulAssert(dupReachable(_depth), "Stack too deep");
+		m_data->push_back((*m_data)[depthToOffset(_depth).value]);
+		if constexpr (!std::is_same_v<Callbacks, NoOpStackManipulationCallbacks>)
+			m_callbacks.dup(_depth.value + 1);
+	}
+	void dup(Offset const& _offset)	{ dup(offsetToDepth(_offset)); }
+
+	bool dupReachable(Offset const& _offset) const noexcept { return dupReachable(offsetToDepth(_offset)); }
+	bool dupReachable(Depth const& _depth) const noexcept { return _depth < size() && 1 <= _depth.value + 1 && _depth.value + 1 <= reachableStackDepth; }
```

Comment:
```suggestion
	bool dupReachable(Depth const& _depth) const noexcept { return _depth < size() && _depth.value + 1 <= reachableStackDepth; }
```

`_depth.value` is `size_t`, isn't it always `>= 0`?

Reply:
yeah hehe